[SPARK-40034][SQL] PathOutputCommitters to support dynamic partitions

steveloughran opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


Uses the StreamCapabilities probe in MAPREDUCE-7403 to identify when a
PathOutputCommitter is compatible with dynamic partition overwrite.

This patch has unit tests but not integration tests; really needs
to test the SQL commands through the manifest committer into gcs/abfs,
or at least local fs. That would be possible once hadoop 3.3.5 is out...

Uses the StreamCapabilities probe in MAPREDUCE-7403 to identify when a
PathOutputCommitter is compatible with dynamic partition overwrite.


### Why are the changes needed?

Hadoop 3.3.5 adds a new committer in mapreduce-core which works fast and correctly on azure and gcs. (it would also work on hdfs, but its optimised for the cloud stores).

The stores and the committer do meet the requirements of Spark SQL Dynamic Partition Overwrite, so it is OK to for spark to work through it.

Spark does not know this; MAPREDUCE-7403 adds a way for any PathOutputCommitter to declare that they are compatible; the IntermediateManifestCommitter will do so.
(https://github.com/apache/hadoop/pull/4728)

### Does this PR introduce _any_ user-facing change?

No.

There is documentation on the feature in the hadoop [manifest committer](https://github.com/apache/hadoop/blob/82372d0d22e696643ad97490bc902fb6d17a6382/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/site/markdown/manifest_committer.md) docs.

### How was this patch tested?

1. Unit tests in hadoop-cloud which work with hadoop versions with/without the matching change.
2. New integration tests in https://github.com/hortonworks-spark/cloud-integration which require spark to be built against hadoop with the manifest committer declaring compatibility

Those new integration tests include

* spark sql test derived from spark's own [CloudRelationBasicSuite.scala#L212](https://github.com/hortonworks-spark/cloud-integration/blob/master/cloud-examples/src/test/scala/org/apache/spark/sql/sources/CloudRelationBasicSuite.scala#L212)
* Dataset tests extended to verify support for/rejection of dynamic partition overwrite [AbstractCommitDataframeSuite.scala#L151](https://github.com/hortonworks-spark/cloud-integration/blob/master/cloud-examples/src/test/scala/com/cloudera/spark/cloud/committers/AbstractCommitDataframeSuite.scala#L151)

Tested against azure cardiff with the manifest committer; s3 london (s3a committers reject dynamic partition overwrites)

steveloughran committed:
SPARK-40034. PathOutputCommitters to support dynamic partition overwrite.

Uses the StreamCapabilities probe in MAPREDUCE-7403 to identify when a
PathOutputCommitter is compatible with dynamic partition overwrite.

This patch has unit tests but not integration tests; really needs
to test the SQL commands through the manifest committer into gcs/abfs,
or at least local fs. That would be possible once hadoop 3.3.5 is out...
Change-Id: I5cbc391bc021b4dd177374e82de9fc33137ac319

Change-Id: I772caf861d6c92f0da6d9a02d9f899236ddaddf9

steveloughran committed:
SPARK-40034. reject newTaskTempFileAbsPath() when dynamic partition unsupported

I believe this was always implicit; only committers with dynamic partition
overwrite would be asked for absolute path temp files(*). With this change
it is explicit, with tests.

(*) certainly nobody has ever complained about it not working with
the s3a committers

Change-Id: I57c2a02ad799f7ab5d9d0a3053da24f960bad289

steveloughran committed:
SPARK-40034. skip parquet dynamic binding tests on incompatible hadoop versions

If mapreduce-core BindingPathOutputCommitter doesn't implement StreamCapabilities
the probes for dynamic commit support through the parquet committer
don't work. So skip that bit of the test case

Change-Id: I5225c70a54c63adf858a9f429fddad251b79783e

steveloughran commented:
this should interest @sunchao and @dongjoon-hyun -know that this doesn't add support to the s3a committers; s3 itself doesn't do the right thing (rename()). Does for abfs and gcs through the manifest committer.

attilapiros commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -140,6 +162,28 @@ class PathOutputCommitProtocol(
+   * @return a path string
+   * @throws UnsupportedOperationException if incompatible
+   */
+  override def newTaskTempFileAbsPath(
+    taskContext: TaskAttemptContext,
+    absoluteDir: String,
+    spec: FileNameSpec): String =  {
+
+    if (supportsDynamicPartitions) {
+      super.newTaskTempFileAbsPath(taskContext, absoluteDir, spec)
+    } else {
+      throw new UnsupportedOperationException(s"Absolute output locations not supported" +
+        s" by committer $committer")
+    }
+  }
```
Do we really need this? I mean in the `setupCommitter()` we already checked the `supportsDynamicPartitions` and an exception was thrown.

steveloughran replied:
adding a safety defence.

attilapiros commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -51,14 +51,7 @@ class PathOutputCommitProtocol(
-    // choice to handle partitioning.
-    throw new IOException(PathOutputCommitProtocol.UNSUPPORTED)
-  }
+  extends HadoopMapReduceCommitProtocol(jobId, dest, dynamicPartitionOverwrite) with Serializable {
```
Please update the comment above the class:
> dynamic partition overwrite is not supported, so that committers for stores which do not support rename will not get confused.

steveloughran replied:
added an explanation of what now happens

attilapiros reviewed (commented):
Mostly a bunch of Nits so far.

steveloughran committed:
SPARK-40034. address review feedback

Change-Id: I6ddc92f56d8762cebb76857a30a5b9dd4fe4948d

steveloughran committed:
SPARK-40034. Documentation changes

* Section in cloud-integration docs
* Add references to committers and papers related to them.
* Remove hadoop openstack reference (it's going to be cut soon)

No mention of the Intermediate Manifest Committer until it is shipped
in an ASF release. It is in Cloudera CDH and has been trouble free,
unlike FileOutputCommitter with abfs (scale) and gcs (correctness).

Change-Id: I97bf56336f6fd6cbd6d56e87c911e62a6deff9c8

steveloughran commented:
the hadoop side of this change is now merged in.

@attilapiros do you have any time to review this again?

attilapiros commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -114,10 +112,33 @@ class PathOutputCommitProtocol(
+   * @return true if the committer declares itself compatible.
+   */
+  private def supportsDynamicPartitions = {
+    committer.isInstanceOf[FileOutputCommitter] ||
```
Do we change the old behaviour because of  the `committer.isInstanceOf[FileOutputCommitter]` part of the condition? As before this PR we were simply throwing an exception.
If this is a change then the `Does this PR introduce any user-facing change?` section of the PR description must be updated too.

steveloughran replied:
up until now, if you used file output committer through the factory, it would reject the use of dynamic partitions. now it lets the caller without needing to implement the path capability.

this does ensure that you can use that committer through the factory mechanism even on hadoop versions <= 3.3.4. I didn't add the path capabilities probe to the original FileOutputCommitter because I'm scared of going near that file: old, complex, critical code

+ will update the user facing details

attilapiros commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -76,6 +78,16 @@ class CommitterBindingSuite extends SparkFunSuite {
     assert(inner.jobAborted, s"$inner job not aborted")
+
+    val binding = new BindingPathOutputCommitter(path, tContext)
+    if (binding.isInstanceOf[StreamCapabilities]) {
```
Now the test is in the `hadoop-3` directory so do we still need this `if`?

steveloughran replied:
yes, because the new feature isn't in hadoop 3.3.4 or earlier....this test suite works with the current and future hadoop releases. the new feature needs the forthcoming hadoop branch-3.3. release

attilapiros replied:
I see.

attilapiros commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -161,7 +204,16 @@ object PathOutputCommitProtocol {
+  /**
+   * Scheme prefix for per-filesystem scheme committers.
+   */
+  private[cloud] val OUTPUTCOMMITTER_FACTORY_SCHEME = "mapreduce.outputcommitter.factory.scheme"
```
Our documentation lack mentioning the `mapreduce.outputcommitter.factory.scheme` config and as I see this config will be more important than it was before (when it was just simply set to a single specific value in `SparkContext#fillMissingMagicCommitterConfsIfNeeded`). 

What about extending the `cloud-integration.md` with some info about its role? Or just add a link to the Hadoop's documentation of this config?

steveloughran replied:
linked to the hadoop docs. we don't yet have a release with the manifest committer docs, but it is covered in https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/site/markdown/manifest_committer.md

attilapiros commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -125,21 +132,57 @@ class CommitterBindingSuite extends SparkFunSuite {
+    }
+  }
+
+  test("permit dynamic partitioning if the committer says it works") {
```
Could you please add a new test where the exact same things are checked just using  `FileOutputCommitterFactory` at the binding (by extracting the test body into separate method(s) and reusing most of the parts)?

steveloughran replied:
done

steveloughran committed:
SPARK-40034. review feedback

1. docs
2. tests

Change-Id: Ia31cf91999157057f1a85061826da74db7f1713e

steveloughran commented:
added a new test case and updated the docs. I've not yet rebased it/merged it with your committer work, but the docs shouldn't clash. once this PR is in I will set my local build up to run your tests against s3 london

attilapiros reviewed (approved):
Some checkstyle issues:
- error file=/Users/attilazsoltpiros/git/attilapiros/spark-review/hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala message=File line length exceeds 100 characters line=213
- error file=/Users/attilazsoltpiros/git/attilapiros/spark-review/hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala message=No space after token = line=177 column=32
- error file=/Users/attilazsoltpiros/git/attilapiros/spark-review/hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala message=File line length exceeds 100 characters line=202

Otherwise LGTM.

dongjoon-hyun commented on docs/cloud-integration.md:
```diff
@@ -266,26 +271,123 @@ It has been tested with the most common formats supported by Spark.
+This deliviers performance and scalability on the object stores.
+
+It is not critical for job correctness to use this with Azure storage; the
+classic FileOutputCommitter is safe there -however this new committer scales
```
nit. `-however` -> `- however`

dongjoon-hyun commented on docs/cloud-integration.md:
```diff
@@ -266,26 +271,123 @@ It has been tested with the most common formats supported by Spark.
+classic FileOutputCommitter is safe there -however this new committer scales
+better for large jobs with deep and wide directory trees.
+
+Because google GCS does not support atomic directory renaming,
```
`google` -> `Google`.

dongjoon-hyun commented on hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/PathOutputCommitProtocol.scala:
```diff
@@ -161,7 +204,16 @@ object PathOutputCommitProtocol {
     " dynamicPartitionOverwrite"
 
+  /**
+     * Stream Capabilities probe for spark dynamic partitioning compatibility.
```
indentation?

steveloughran replied:
fixed. interesting. i've switched my IDE to rendering scala/javadocs and it hid that. something for me to worry about. maybe I should turn it off again

dongjoon-hyun commented on hadoop-cloud/src/hadoop-3/test/scala/org/apache/spark/internal/io/cloud/CommitterBindingSuite.scala:
```diff
@@ -125,22 +134,126 @@ class CommitterBindingSuite extends SparkFunSuite {
-
     val protocol = instance.asInstanceOf[PathOutputCommitProtocol]
     assert("file:///tmp" === protocol.destination)
   }
```
If you don't mind, could you revert the newline addition on the above test case, `test("committer protocol can be serialized and deserialized")`?

steveloughran replied:
done. went and compared the history and reverted any newline changes which expanded the diff

dongjoon-hyun reviewed (approved):
cc @sunchao, @viirya , @cloud-fan

steveloughran committed:
SPARK-40034. review feedback

Change-Id: I71a15ba3909a8912351987ad2dfbba8dca83b5b8

dongjoon-hyun commented:
BTW, when is the ETA for Apache Hadoop 3.3.5, @steveloughran ?

steveloughran commented:
@dongjoon-hyun I'm off on vacation next week; we will fork off the branch the week after.

things i'd like in *if anyone has the time*
1. upgraded shaded parquet
2. the shaded avro PR
3. get that arm64 docker image working for a release there

dongjoon-hyun commented:
Merged to master for Apache Spark 3.4.0.

dongjoon-hyun commented:
Thank you, @steveloughran and @attilapiros .

dongjoon-hyun commented:
@attilapiros FYI, there is a new PR for this area.
- https://github.com/apache/spark/pull/40221

dongjoon-hyun commented:
This is reverted from `branch-3.4` only while being kept in `master` branch for Apache Spark 3.5.
- https://github.com/apache/spark/commit/57aa3d1ca6e17e4c6b934d74176ea22ca56d60f7
